### ISSUE TITLE
parallel netcdf4

### ISSUE THREAD

**Issue by apatlpo** [spectralDNS/shenfun#9]
It'd be great to write netcdf output files. 
This is the standard for many users in the geophysical community.
Parallel file writing is implemented in [netCDF4](http://unidata.github.io/netcdf4-python/).
I failed to install it via conda however.
Could it be possible to have it in the spectralDNS channel as it is the case for hdf5-parallel?

**Comment by mikaem** [comment]
Hi
I have not used netcdf4, but I would not object to someone adding support for it. I see that netcdf4 is available on [conda-forge](https://github.com/conda-forge/netcdf4-feedstock), but this is only serial version of the library. And there's a [libnetcdf](https://github.com/conda-forge/libnetcdf-feedstock). Are both required? Should probably only be an issue of using these recipes with hdf5-parallel instead of the serial versions used on conda-forge I think.

**Comment by mikaem** [comment]
It built quite easily on Ubuntu, so I put two versions, libnetcdf-parallel and netcdf4-parallel on Anaconda cloud. Feel free to check them out. The last one was built with python=3.6. Let me know if you need anything else.

**Comment by apatlpo** [comment]
That's great, thanks, but today I'm on osx unfortunately.
Would you be able at some point to add an osx version of these libraries?

**Comment by mikaem** [comment]
I'll see what I can do.

**Comment by mikaem** [comment]
@apatlpo I have now uploaded netcdf built for osx to the spectralDNS channel.

**Comment by apatlpo** [comment]
Thanks it works like a charm for me !
Starting from the hdf5 file writer, I wrote a [netcdf file writer](https://github.com/apatlpo/shenfun_trial/blob/master/swater_1L/nc_writer.py).
It works well for a [simple configuration](https://github.com/apatlpo/shenfun_trial/blob/master/swater_1L/swater_1L_physical.py).
Let me know if you want me to push a pull request to shenfun of this file.

**Comment by mikaem** [comment]
Great:-) Go ahead with the pull request. Perhaps you can even improve it? I have spent way too little time on the IO-interface.

**Comment by apatlpo** [comment]
ok, will proceed with the PR ( beware this will be my first :) ).

As far as improvements, I don't have time to do it right now but here is a to do list for the future:

- [ ] Add slicing capability
- [ ] Optional names for coordinates
- [ ] Optional number of time steps per file

**Comment by apatlpo** [comment]
I just reinstalled shenfun via:
```
conda build -c conda-forge -c spectralDNS conf/conda
conda install -c conda-forge -c spectralDNS shenfun --use-local
```
This does not install netCDF4 which raises the following warning at imports:
```
>>> import shenfun
/Users/aponte/.miniconda3/envs/shenfun/lib/python3.6/site-packages/shenfun/utilities/nc_writer.py:11: UserWarning: netcdf not installed
  warnings.warn('netcdf not installed')
```
Is this intended?

As a side note, it may be nice to suggest the creation of a shenfun conda environment prior to running
these last install lines in the doc.

**Comment by mikaem** [comment]
Not intended at all. I see that netcdf is not added to [meta.yaml](https://github.com/spectralDNS/shenfun/blob/master/conf/conda/meta.yaml). Do you remember what the requirements were?

And you are correct about the last note. I'll update the readme.

**Comment by mikaem** [comment]
I think it should be netcdf4-parallel and libnetcdf-parallel, right?

**Comment by mikaem** [comment]
And there should be some tests on the netcdf io...

**Comment by apatlpo** [comment]
Great thanks.

Regarding tests, should I just propose additions to 'test/test_io.py' ?

**Comment by mikaem** [comment]
Yes, that would be great. 

I have written a little bit about post-processing in the [documentation](http://shenfun.readthedocs.io/en/latest/gettingstarted.html#post-processing). Have not written much about netcdf and one reason for this is that it does not have the option to store slices. Are you not interested in this feature?

**Comment by apatlpo** [comment]
just letting you know I am not dropping this.

I am fighting a `ModuleNotFoundError: No module named 'mpi4py_fft` at the moment.

**Comment by mikaem** [comment]
That means you haven't installed the [mpi4py-fft](https://anaconda.org/spectralDNS/mpi4py_fft) package. Should be automatically installed if you used conda...

**Comment by apatlpo** [comment]
Yes, I am using conda.
I have restarted from scratch the environment.
I don't understand what is going on.

**Comment by apatlpo** [comment]
Where is the `mpi4py-fft` dependency explicited in the library?

**Comment by apatlpo** [comment]
Shouldn't it be added to `conf/conda/meta.yaml`?

**Comment by mikaem** [comment]
It's in the components-requirements file

**Comment by mikaem** [comment]
No. It is only available as source for now. There's no package available anywhere so meta.yaml won't work.

**Comment by apatlpo** [comment]
nevermind, I got confused because it is not showing up with `conda list`.
shenfun is up and running. 
sorry for the noise

**Comment by mikaem** [comment]
puh:-)

**Comment by mikaem** [comment]
But you should be able to install it from the spectralDNS channel, right? I forgot it was there as well.

**Comment by apatlpo** [comment]
I followed:
```
conda build -c conda-forge -c spectralDNS conf/conda
conda create --name shenfun -c conda-forge -c spectralDNS shenfun --use-local
conda activate shenfun
```

I've added netcdf tests in `test_io.py` but am at a loss as what is the best to check if the test
passes.
What would you do ?

**Comment by mikaem** [comment]
I don't follow. You mean what is the best way to call the test itself? like `py.test -v test_io.py`?

**Comment by apatlpo** [comment]
yep, sorry I have never done this before ...
(note for myself, this was required: `pip install -U pytest`)


Tests seem to pass. I'll send a PR soon.

Regarding slices, I am indeed interested in that and do not see any particular reasons for why one could not slice with netcdf.
